### DISCUSSION TITLE
Does Airflow request metadata database postgreSQL through persistent connections?

### DISCUSSION THREAD

**Discussion by kevinchcn** [apache/airflow#33644] [Q&A] [answered]
We want to use it in a production environment because there are safety requirements, so we hope to receive a confident answer.
If so, can I use a short connection?

Thanks a lot.

**Comment by potiuk** [comment]
Could you please elaborate what do you mean by persistent connections and what do you want to achieve? It's hard to guess what you are asking about so, I could **confidently** answer `yes, for sure` but I am absolutely not sure if my answer will be answering your question.

**Comment by potiuk** [comment]
The best way to get confident answer is to ask very clear and precise question that can be easily answered yes/no.

**Comment by kevinchcn** [comment]
Thanks for your reply @potiuk
ok, let me try to explain.

Persistent connections means that after Airflow connects to postgresql for the first time to complete reading and writing, it does not actively close the connection and does not perform any operations, but it will still continue to connect for a few days, months, or even all the time.

What I want to achieve?

1. If the Postgresql is down, will Airflow also be down?
2. If the database restarts after one hour of downtime, can Airflow automatically reconnect and provide services.

The question I get is
```
The Airflow uses persistent connections?
As a global standard, our firewall terminates persistent connections inactive for more than 1 hour. In order to avoid unexpected impact on the Airflow, please use short connections, or enable heartbeat or re-connection mechanism for the Airflow.

```

Thanks.

**Comment by potiuk** [comment, marked as answer]
1. If the Postgresql is down, will Airflow also be down?

yes

2. If the database restarts after one hour of downtime, can Airflow automatically reconnect and provide services.

depends. usually airflow will go down when postgres is down. then your deployment will (hopefully) restart it and connection will be closed

> Persistent connections means that after Airflow connects to postgresql for the first time to complete reading and writing, it does not actively close the connection and does not perform any operations, but it will still continue to connect for a few days, months, or even all the time.

This is a strange definition, I wonder where you took it from. Generally Airflow is a distributed system and it has many components - potentially every component opens a separate connection to the DB. Typically (look it up in our docs) there should be - for example for Postgres - PG bouncer between airflow and DB, so while airflow components might open multuple dynamic connections (even hundreds of them - depends how many tasks you have running at a given time, the PG bouncer manages multiplexing of those connections so that there is just a handful of those connections opened at a time. And depending on which components are restarting, the behaviours might cchange - for exmple when PG bouncer is restarting the DB connections might get broken,but when just scheduler, or just worker, or just webserver get restarted only connections from them to pgbouncer will get down but PGBouncer connections to the DB will go down.

So there are many scenarios to consider.

> As a global standard, our firewall terminates persistent connections inactive for more than 1 hour. In order to avoid unexpected impact on the Airflow, please use short connections, or enable heartbeat or re-connection mechanism for the Airflow.

This is somethibg that you can consider on a driver level. Each database has it's own way to do heartbeating and if you wish to have heattbetating enabled you should look at the capabilities of driver you want to configure. Airlflow uses sqlalchemy and allows to configure various parameters of the DB connection for example - engine args:  https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#sql-alchemy-engine-args - and you need to make sure that you set the args of the DB you are using in the way that your firewall can handle.

We did not want to make an opinionated decisions on those so Airfflow allows Deployment manager to make the right set of choices and choose the right drivers that can enable those kind of features - it's up to the Deployment manager to choose the right configuration/choose the driver that gives heartbeating capabilities.

I hope you will find it helpful.

**Reply by kevinchcn** [reply]
Thanks @potiuk
yes it's very helpful.

I need to spend some time and do some tests to digest your answer.
anyway, thank u very much!

> Persistent connections means that after Airflow connects to postgresql for the first time to complete reading and writing, it does not actively close the connection and does not perform any operations, but it will still continue to connect for a few days, months, or even all the time.
This definition was modified based on long connections and may not be accurate. Please ignore it
